View server logs

rcrichton opened the pull request:
Closes #334

@devcritter could you please review this for me?

@BMartinos, you may want to have a look as well if you have space, otherwise don't worry.

![server-logs](https://cloud.githubusercontent.com/assets/266576/10905081/9cc65c16-8220-11e5-9123-c5a583645651.gif)

devcritter commented on app/index.html:
```diff
@@ -144,6 +146,8 @@
     <script src="bower_components/angular-highlightjs/build/angular-highlightjs.js"></script>
     <script src="bower_components/ng-file-upload/angular-file-upload.js"></script>
     <script src="bower_components/ng-file-upload-shim/angular-file-upload-shim.js"></script>
+    <script src="bower_components/eonasdan-bootstrap-datetimepicker/build/js/bootstrap-datetimepicker.min.js"></script>
```
I take you had issues with the existing datepicker? If so, I wonder if we shouldn't swap out the existing ones for the new one so as to make it consistent?

cc @BMartinos

BMartinos replied:
Initially time based filtering wasn't needed for transactions/audits but I agree with @devcritter that keeping it consistent will be better. We might just need to check but I cant see any reason why swapping the old one with the new one shouldn't return the same result. The date that will be sent to the API will just contain extra "time" information

rcrichton replied:
I didn't have issues with the existing datepicker but I needed a date**time**picker instead. So I had to find another one.

devcritter replied:
They've got different stylings though. Does the new one have an option to use a datepicker (no -time)? Tbh I'm not so crazy about them co-existing due to the inconsistent styling.

rcrichton replied:
Hmm, I see what you mean. I thought they were similarly styled but I see its pretty different. I agree then, we should change them all, I'm sure we can disable the time component of the new one. It's pretty configurable. I do think this should be done in a separate issue thought. It doesn't fit here. We could create an issue and make sure it gets done before the next release? How does that sound?

devcritter replied:
Yeah, sounds good

rcrichton replied:
Ok here is the new issue: https://github.com/jembi/openhim-console/issues/338

devcritter commented on app/scripts/controllers/logs.js:
```diff
@@ -0,0 +1,96 @@
+angular.module('openhimConsoleApp')
+  .controller('LogsCtrl', function ($scope, $location, Api) {
+    function formatLog(log) {
+      return new moment(log.timestamp).format('YYYY-MM-D HH:mm:ss.SSS') + ' - ' + log.level + ': [' + log.label + '] ' + log.message + '\n';
```
Only one D in the date format

rcrichton replied:
Will fix.

devcritter commented on app/scripts/controllers/logs.js:
```diff
@@ -0,0 +1,96 @@
+      lastFetch = now;
+      Api.Logs.query(localParam, function (results) {
+        results.forEach(function (log) {
+          $scope.logs += formatLog(log);
```
It might be good to enforce a max size here - maybe keep a line count if it's above 100k then start removing from the front?

rcrichton replied:
The core API is limited to 100 000 lines at a maximum, do you think we should limit here as well?

devcritter replied:
I was thinking that if you leave the page open and let it keep auto appending it might make that string really big after a while - although to be fair I'm not sure if it'll actually be a problem in practice.

devcritter commented on app/scripts/controllers/logs.js:
```diff
@@ -0,0 +1,96 @@
+      delete $scope.params.from;
+      delete $scope.params.until;
+      $scope.params.level = 'info';
+      $scope.reload();
```
Atm when you hit 'clear' on the interface it keeps the existing data, which is a bit annoying when the logs have filled up a lot. It would be good to set $scope.logs to empty here, I guess before calling reload.

rcrichton replied:
Hmm, it should be clearing the logs when it reloads the controller. Is this consistently not working for you?

BMartinos replied:
Not sure if this could be the issue but if you trying to reload the controller by setting the location.search paramaters to the same parameters then the reload doesn't happen. Angular ignores this then as the parameters are the same

rcrichton replied:
Right, I think that is what you are seeing @devcritter. I'm not really sure this is an issue as the clear button is intended to reset the filters. Would changing the button text from 'Clear' to 'Reset' be better?

devcritter replied:
Aaah ok I see - cool yeah; or maybe 'Clear Filters'? (That's what it's currently called on the transactions page.)

rcrichton replied:
Clear Filters may be best of consistency. I'l update.

BMartinos commented:
Hey @rcrichton I havent looked to much into the code yet but from looking at the awesome gif :) i think the UI/UX is looking great! Just one thing i noticed is that if you click on a date it automatically adds in the current time ( 13:45 ), Would it be possible to have this as 00:00 and then give the user the choice to specify the time if they need it? 

Otherwise, i think it looks great!

devcritter commented on app/scripts/controllers/logs.js:
```diff
@@ -0,0 +1,96 @@
+      $scope.autoupdate = false;
+      $scope.tailLogs = false;
+    } else {
+      $scope.autoupdate = true;
```
Can you add a control to toggle autoupdate on the interface? It would be useful if you want to study the logs without having it consistently updating. Another idea could be to merge tail and autoupdate? I.e. only autoupdate when tail is true?

rcrichton replied:
I could but I I didn't want to give the user too much choice. Also, autoupdate only functions when there isn't a date range set so would be disabled a lot of the time. Also with tailing off its easy to keep you place as logs update automatically as your view won't be reset. I don't mind changing it, but I'm happy with how it works atm. So, let me know how strongly you feel about this.

devcritter commented:
@rcrichton looks good so far - just some comments above.

rcrichton commented:
Cool, thanks guys.

@BMartinos I agree 0'ing the time would be better. I'l see if that can be done.

rcrichton commented:
I've pushed some changes to address the comments.

* The time now 0's
* Changed 'Clear' to 'Reset' - is this better?
* I left the autoupdate thing, see comments above.
* I added some code to keep the logs string a reasonable size.
* Fixed a date format
* Added an issue for the changing the other datepickers to the new one.

Let me know what you think @devcritter?

devcritter commented:
@rcrichton Cool looks good! :+1: The '0' time is really nice now. No objections if you want to leave the autoupdate.
